// @password-validator/core/dist/cjs/password/api/standard/PasswordValidatorManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PasswordValidatorManager = void 0;
const ValidationResult_js_1 = require("../../validator/ValidationResult.js");
const FluentPasswordValidator_js_1 = require("../fluent/FluentPasswordValidator.js");
const PasswordValidatorConflictException_js_1 = require("../../validator/PasswordValidatorConflictException.js");
class PasswordValidatorManager {
    registeredValidators = Array.of();
    constructor() { }
    static standard() {
        return new PasswordValidatorManager();
    }
    static fluent() {
        return new FluentPasswordValidator_js_1.FluentPasswordValidator();
    }
    register(...validators) {
        const conflictMessages = Array.of();
        this.verifyEachValidatorAgainstThemselves(validators, conflictMessages);
        for (const validator of validators) {
            this.verifyAgainstOtherValidators(validator, conflictMessages);
            this.registerValidatorWithNoConflict(conflictMessages, validator);
        }
        this.handleConflicts(conflictMessages);
    }
    verifyEachValidatorAgainstThemselves(validators, conflictMessages) {
        for (let i = 0; i < validators.length; i++) {
            for (let j = i + 1; j < validators.length; j++) {
                const conflictMsg = validators[i].conflictsWith(validators[j]);
                if (conflictMsg.length > 0) {
                    conflictMessages.push(...conflictMsg);
                }
            }
        }
    }
    verifyAgainstOtherValidators(validator, conflictMessages) {
        for (const registeredValidator of this.registeredValidators) {
            const conflictMsg = validator.conflictsWith(registeredValidator);
            if (this.hasConflict(conflictMsg)) {
                conflictMessages.push(...conflictMsg);
                break;
            }
        }
    }
    registerValidatorWithNoConflict(conflictMessages, validator) {
        if (!this.hasConflict(conflictMessages)) {
            this.registeredValidators.push(validator);
        }
    }
    handleConflicts(conflictMessages) {
        if (this.hasConflict(conflictMessages)) {
            let message = "";
            let errorCount = 0;
            for (const conflictMessage of conflictMessages) {
                errorCount++;
                message += `${errorCount}: ${conflictMessage}\n`;
            }
            throw new PasswordValidatorConflictException_js_1.PasswordValidatorConflictException(message);
        }
    }
    hasConflict(conflictMsg) {
        return conflictMsg.length > 0;
    }
    validators() {
        return this.registeredValidators;
    }
    validate(password) {
        const messages = Array.of();
        let isValid = true;
        for (const validator of this.registeredValidators) {
            const result = validator.validate(password);
            if (result.isValid()) {
                continue;
            }
            isValid = false;
            messages.push(...result.getMessages());
        }
        return new ValidationResult_js_1.ValidationResult(isValid, messages);
    }
}
exports.PasswordValidatorManager = PasswordValidatorManager;
