// @password-validator/core/dist/cjs/password/api/fluent/FluentPasswordValidator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FluentPasswordValidator = void 0;
const DigitValidator_js_1 = require("../../validator/DigitValidator.js");
const UpperCaseValidator_js_1 = require("../../validator/UpperCaseValidator.js");
const MinLengthValidator_js_1 = require("../../validator/MinLengthValidator.js");
const MaxLengthValidator_js_1 = require("../../validator/MaxLengthValidator.js");
const LowerCaseValidator_js_1 = require("../../validator/LowerCaseValidator.js");
const PasswordValidatorManager_js_1 = require("../standard/PasswordValidatorManager.js");
const NoSpaceCharacterValidator_js_1 = require("../../validator/NoSpaceCharacterValidator.js");
const SpecialCharacterValidator_js_1 = require("../../validator/SpecialCharacterValidator.js");
class FluentPasswordValidator {
    digitValidator;
    lowerCaseValidator;
    minLengthValidator;
    maxLengthValidator;
    upperCaseValidator;
    noSpaceCharacterValidator;
    specialCharacterValidator;
    min(passwordRule) {
        this.minLengthValidator = new MinLengthValidator_js_1.MinLengthValidator(passwordRule);
        return this;
    }
    max(passwordRule) {
        this.maxLengthValidator = new MaxLengthValidator_js_1.MaxLengthValidator(passwordRule);
        return this;
    }
    lower(passwordRule) {
        this.lowerCaseValidator = new LowerCaseValidator_js_1.LowerCaseValidator(passwordRule);
        return this;
    }
    upper(passwordRule) {
        this.upperCaseValidator = new UpperCaseValidator_js_1.UpperCaseValidator(passwordRule);
        return this;
    }
    digit(passwordRule) {
        this.digitValidator = new DigitValidator_js_1.DigitValidator(passwordRule);
        return this;
    }
    noSpace() {
        this.noSpaceCharacterValidator = new NoSpaceCharacterValidator_js_1.NoSpaceCharacterValidator();
        return this;
    }
    specialCharacter(passwordRule) {
        this.specialCharacterValidator = new SpecialCharacterValidator_js_1.SpecialCharacterValidator(passwordRule);
        return this;
    }
    validate(password) {
        const validators = [
            this.digitValidator,
            this.minLengthValidator,
            this.maxLengthValidator,
            this.lowerCaseValidator,
            this.upperCaseValidator,
            this.noSpaceCharacterValidator,
            this.specialCharacterValidator,
        ];
        const pm = PasswordValidatorManager_js_1.PasswordValidatorManager.standard();
        for (const validator of validators) {
            if (!validator) {
                continue;
            }
            pm.register(validator);
        }
        return pm.validate(password);
    }
}
exports.FluentPasswordValidator = FluentPasswordValidator;
