// @password-validator/core/dist/esm/password/api/fluent/FluentPasswordValidator.js
import { DigitValidator } from "../../validator/DigitValidator.js";
import { UpperCaseValidator } from "../../validator/UpperCaseValidator.js";
import { MinLengthValidator } from "../../validator/MinLengthValidator.js";
import { MaxLengthValidator } from "../../validator/MaxLengthValidator.js";
import { LowerCaseValidator } from "../../validator/LowerCaseValidator.js";
import { PasswordValidatorManager } from "../standard/PasswordValidatorManager.js";
import { NoSpaceCharacterValidator } from "../../validator/NoSpaceCharacterValidator.js";
import { SpecialCharacterValidator } from "../../validator/SpecialCharacterValidator.js";
export class FluentPasswordValidator {
    digitValidator;
    lowerCaseValidator;
    minLengthValidator;
    maxLengthValidator;
    upperCaseValidator;
    noSpaceCharacterValidator;
    specialCharacterValidator;
    min(passwordRule) {
        this.minLengthValidator = new MinLengthValidator(passwordRule);
        return this;
    }
    max(passwordRule) {
        this.maxLengthValidator = new MaxLengthValidator(passwordRule);
        return this;
    }
    lower(passwordRule) {
        this.lowerCaseValidator = new LowerCaseValidator(passwordRule);
        return this;
    }
    upper(passwordRule) {
        this.upperCaseValidator = new UpperCaseValidator(passwordRule);
        return this;
    }
    digit(passwordRule) {
        this.digitValidator = new DigitValidator(passwordRule);
        return this;
    }
    noSpace() {
        this.noSpaceCharacterValidator = new NoSpaceCharacterValidator();
        return this;
    }
    specialCharacter(passwordRule) {
        this.specialCharacterValidator = new SpecialCharacterValidator(passwordRule);
        return this;
    }
    validate(password) {
        const validators = [
            this.digitValidator,
            this.minLengthValidator,
            this.maxLengthValidator,
            this.lowerCaseValidator,
            this.upperCaseValidator,
            this.noSpaceCharacterValidator,
            this.specialCharacterValidator,
        ];
        const pm = PasswordValidatorManager.standard();
        for (const validator of validators) {
            if (!validator) {
                continue;
            }
            pm.register(validator);
        }
        return pm.validate(password);
    }
}
